src/irmin/tree.ml

```diff
@@ -1585,6 +1879,16 @@ module Make (P : Private.S) = struct
     in
     tree t (fun x -> Lwt.return x)
 
+  let key (t : t) =
+    Log.debug (fun l -> l "Tree.key");
+    match t with
+    | `Node n -> (
+        match Node.key n with Some key -> Some (`Node key) | None -> None)
+    | `Contents (c, m) -> (
+        match Contents.key c with
+        | Some key -> Some (`Contents (key, m))
+        | None -> None)
```

Comment:
`Store.Tree.key` returns `None` if the key is not cached, even if the key could be fetched from the backend using `index`. Is that behaviour intentional? 

For example, `Store.Tree.key b` is `None` in:
```ml
let a = Store.Commit.tree my_valid_commit
(* a exists in the backend *)

let b = a |> Store.Tree.to_concrete |> Store.Tree.of_concrete
(* b exists in the backend but key is lost *)
```

Reply:
The existing semantics are intended (and reflect that there's no hash-only shallowed tree in the API), but I think the other use-case is valid. We could add e.g. `find_key : t -> kinded_key option Lwt.t`, or just count on the user doing `Tree.hash` and then a manual index. WDYT?

Reply:
I like this `find_key` solution but I have the strange feeling that the `key` function could be renamed to `cached_key`

Reply:
I'd rather avoid `cached_key` for two reasons:

- it's not the case that objects constructed with `Store.Tree.of_concrete` have "uncached" keys: they simply don't have keys at all, because they're not references to some store position. It's possible that some objects with the same _hash_ have keys in the store, but there may be zero-or-more of those (2+ in the case of a partial index).

- objects constructed with `Store.Tree.shallow` don't really have "cached" keys: it's not possible to discard the key and still have a usable tree (and `Tree.clear` won't do this, despite being documented as clearing the caches).

If we were to add `find_key`, I'd interpret it along the lines of "`find_key t` is the key of _an object with the same hash as_ 
 `t` in the underlying store, if such an object is indexed" w/ the distinction being necessary because users can observe duplicate objects if they know the key implementation.

Reply:
I've added `find_key` and improved the documentation of these functions generally.